Allow a single column to be used as rownames in as.matrix

sritchie73 opened the pull request:
Implements #2692 

Added a `rownames` argument to `as.matrix.data.table` and `as.data.frame.data.table` with the following behaviour:

 - `rownames = TRUE` takes `key(x)` as the rownames of the new matrix / data.frame if it is a single column, or the first column if `!haskey(x)` or `length(key(x)) > 1`.
 - `rownames = "column"` takes the named column as the rownames of the new matrix / data.frame.
 - `rownames = 3` looks up the column by index and uses that column as the rownames of the new matrix / data.frame.

Use cases include:

 - Converting a long `data.table` to a matrix via `dcast()`
 - Casting to a `data.frame` for other package's functions that expect data.frame arguments to have rownames for subsetting / row matching in their internal workings.

I've added documentation pages for `as.matrix.data.table` and `as.data.frame.data.table`, describing the use and behaviour of the `rownames()` argument, and highlight examples where it might be useful. The `as.data.frame.data.table` documentation also highlights the `setDF` function for cases where rownames nor a copy of the `data.table` are not required.

I'd appreciate feedback on the amount of error checking implemented (I tend to be overzealous here), and based on that additional unit tests can be implemented.

MichaelChirico commented:
I'd say no need for `as.data.frame.data.table` to be extended -- `setDF` is preferred and has a `rownames` argument already, unless I'm missing something.

sritchie73 commented:
I did see that argument in `setDF`. Does it allow you to use a column in `dt` as the rownames as suggested in #1719 ? Its documentation still suggests you must manually provide the vector yourself.

My understanding was also that `as.data.frame` and `setDF` have two different use cases:

 - `setDF` for when you want to coerce a data.table to a data.frame by reference, and
 - `as.data.frame` when you want to make a copy that is a data.frame.

Otherwise why not just alias `as.data.frame.data.table` to `setDF`? i.e.:

```
as.data.frame.data.table <- function(x, ...) { setDF(x) }
```

MichaelChirico commented:
In fact I'd be fine with `setDF(copy(x))`, which is essentially what's done currently:

```
data.table:::as.data.frame.data.table
function (x, ...) 
{
    ans = copy(x)
    setattr(ans, "row.names", .set_row_names(nrow(x)))
    setattr(ans, "class", "data.frame")
    setattr(ans, "sorted", NULL)
    setattr(ans, ".internal.selfref", NULL)
    ans
}
```

More that I'm advocating for dev time to improve `setDF` over `as.data.frame.data.table`.

Thanks for the PR btw!

sritchie73 commented:
Ah I see, that makes sense.

The tricky part will be trying to remove the column and adding it as the rownames attributes all by reference. I don't see why it couldn't theoretically be done, but might require some C code to implement.

sritchie73 commented:
I've rolled this back so that `rownames` is only implemented for `as.matrix`.

`data.frame`s require more thought and work. I agree the development work should go into `setDF`, and then `as.data.frame` should just call `setDF`. Complicating this however, the `as.data.frame` generic has a `row.names` argument that has the same functionality as `rownames` in `setDF` currently (i.e. a character vector of rownames can be supplied). We'd need to think about how to handle this conflict in arguments.

sritchie73 commented:
A couple  of queries:

 -`rownames` could also accept a vector to use as the rownames of the matrix - Should I add this?
 - I've had to use `with = FALSE` to drop the column containing the rownames, but I understand `with = FALSE` is deprecated (#2620) - is there a better way of doing this?

HughParsonage commented on R/data.table.R:
```diff
@@ -1881,17 +1881,58 @@ chmatch2 <- function(x, table, nomatch=NA_integer_) {
+      # E.g. because rownames is some sort of object that cant be converted to a column index
+      stop("rownames must be TRUE, a column index, or a column name in x")
+    } else {
+      if (is.logical(rownames) && isTRUE(rownames)) {
```
`isTRUE(rownames)` is sufficient?

sritchie73 replied:
You're right. I've changed this statement to `identical(rownames, TRUE)` which I think is clearer (and is used elsewhere in data.table.R).

HughParsonage commented on R/data.table.R:
```diff
@@ -1881,17 +1881,58 @@ chmatch2 <- function(x, table, nomatch=NA_integer_) {
+    rn <- x[[rnc]]
+    dm <- dim(x) - c(0, 1)
+    cn <- names(x)[-rnc]
+    X <- x[, -rnc, with = FALSE]
```
I think `x[, .SD, .SDcols = c(cn)]` or `x[, (rn) := NULL]` could work -- not 100% on what the variables are, or the best way to do this *within* data.table.

sritchie73 commented:
Current check failure appears unrelated to the changes made in my branch

The AppVeyor build is failing in R-devel on test 1774.9:

```
x = rnorm(27)
x[sample(length(x), length(x)/2)] = NA
dim(x) = c(3L,3L,3L)
# na.rm=TRUE / sorted=TRUE
dt = as.data.table(x)
test(1774.9, TRUE, all(
  identical(dim(dt), c(14L,4L)),
  identical(names(dt), c(paste0("V",1:3),"value")),
  dt[is.na(value), .N] == 0L,
  length(key(dt)) > 0L
))
```

None of these commands call `as.matrix.data.table`

mattdowle commented on R/data.table.R:
```diff
@@ -1881,17 +1881,64 @@ chmatch2 <- function(x, table, nomatch=NA_integer_) {
+    rn <- x[[rnc]]
+    dm <- dim(x) - c(0, 1)
+    cn <- names(x)[-rnc]
+    X <- x[, .SD, .SDcols = cn]
```
`x <- x[, -..rnc]` now works here in dev instead of these 2 lines. A copy is needed in this case at this point, iiuc, otherwise, `x[, (rnc):=NULL]`  to remove that column by reference, currently.  And maybe `x[, ..rnc := NULL]` in future.

sritchie73 replied:
Thanks, yes I discovered `-..` was implemented as I was working on this. Ultimately I decided not to use `-..rnc` as you then have to define a dummy `..rnc` at the top of the function to avoid `R CMD check --as-cran` complaining `..rnc` is a global variable with no visible binding.

mattdowle replied:
Interesting, I see.  That is a shame about the no visible binding note.
In general we've tried to use idiomatic data.table internally, and then deal with the no-visible-binding note explicitly by adding a dummy NULL as you say.  This way, when folk look at the internals they see how we use data.table ourselves.  If and when there is ever a solution for the CRAN note, we can fix it one distinct place by taking away the NULL definitions.

MichaelChirico commented on R/data.table.R:
```diff
@@ -1881,17 +1881,64 @@ chmatch2 <- function(x, table, nomatch=NA_integer_) {
-  dm <- dim(x)
-  cn <- names(x)
+as.matrix.data.table <- function(x, rownames, ...) {
+  rn <- NULL
```
not sure we have a style guide on this, but I note that the corresponding CRAN cheat for `[.data.table` symbols are defined in the package environment rather than the function body:

https://github.com/Rdatatable/data.table/blob/master/R/data.table.R#L11

+ side is cutting that tiny bit of overhead for use cases that might repeatedly call this method; - side is increasing potential for unintentional collisions (so if moving outside the body, perhaps use some more obscure name)

jangorecki replied:
they are defined there because they are exported. `rn` won't be used by user.

sritchie73 replied:
Yeap - `rn` is internal here, it will contain the vector of rownames to put in the matrix (after all the processing in `if (!missing(rownames)) {}`. `rnc` will contain the index of the column in `x` to be dropped.

MichaelChirico commented on R/data.table.R:
```diff
@@ -1881,17 +1881,64 @@ chmatch2 <- function(x, table, nomatch=NA_integer_) {
+  if (!missing(rownames)) { # Convert rownames to a column index if possible
+    if (is.null(rownames)) {
+      warning("rownames is NULL, ignoring rownames")
+    } else if (length(rownames) == nrow(x)) {
```
i think `length(rownames)` is called enough times (3) to warrant assigning it, perhaps here as `if (n_rn <- length(rownames))` or analogous

MichaelChirico replied:
It's good that you check vs. `nrow(x)` first for the `nrow(x) == 1L` case; please be sure to document behavior in this case.

MichaelChirico commented on R/data.table.R:
```diff
@@ -1881,17 +1881,64 @@ chmatch2 <- function(x, table, nomatch=NA_integer_) {
+      # rownames argument is a vector of row names, no column in x to drop.
+      rn <- rownames
+      rnc <- NULL 
+    } else if (length(rownames) != 1) {
```
`1L` please :grimacing:

MichaelChirico commented on R/data.table.R:
```diff
@@ -1881,17 +1881,64 @@ chmatch2 <- function(x, table, nomatch=NA_integer_) {
+      stop("rownames must be a single column in x or a vector of row names of length nrow(x)")
+    } else if (is.na(rownames)) {
+      warning("rownames is NA, ignoring rownames")
+    } else if (identical(rownames, FALSE)) {
```
i find it a bit weird that `rownames = TRUE` is accepted, but `rownames = FALSE` is incorrect usage and results in this `warning`. Setting `rownames` dynamically by some condition evaluating to `TRUE` or `FALSE` seems like a natural use case.

sritchie73 replied:
We can drop the warning

sritchie73 commented:
@MichaelChirico can you tell me how to pull your changes to my branch? They're not appearing for me locally so I'm not sure how GitHub works here.

MichaelChirico commented:
you should be able to git pull <name for data.table upstream, usually
either upstream or origin> <name of this branch>

On Wed, Mar 28, 2018, 8:14 AM Scott Ritchie <notifications@github.com>
wrote:

> @MichaelChirico <https://github.com/MichaelChirico> can you tell me how
> to pull your changes to my branch? They're not appearing for me locally so
> I'm not sure how GitHub works here.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/Rdatatable/data.table/pull/2702#issuecomment-376716664>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AHQQdSaXzYU7EWFT59MZaQX4AGY9RYahks5titXcgaJpZM4S4Ite>
> .
>

sritchie73 commented on R/data.table.R:
```diff
@@ -1891,33 +1891,34 @@ as.matrix.data.table <- function(x, rownames, ...) {
-            warning("rownames is TRUE but multiple keys found in key(x), using first column instead")
-            rownames <- 1
+          if (length(rownames) > 1L) {
+            warning(sprintf("rownames is TRUE but multiple keys [%s] found for x; defaulting to first key column [%s]",
```
@MichaelChirico not quite - I've defaulted to just using the first column (rather than the first key column) if there are multiple keys. We can certainly take the first key column if you think its the right approach? (Or perhaps the last key since that is how the rows will be ordered?)

sritchie73 commented:
Thanks @MichaelChirico it turned out `git` was lying to me that my branch was up to date with `origin`. 

I've tried to add comments to your changes but I don't quite understand the GitHub review feature so i'll copy in the comments here:

 - I will drop the warning where `rownames = FALSE`
 - Your change to the warning message where multiple keys are detected is not quite correct - I've defaulted to use `x[,1]` as the rownames in that case rather than `x[,key(x)[1]]`. I can certainly change this – but would think maybe it could be useful to take the last key (since the rows of x will be ordered by that column)?

MichaelChirico commented:
@sritchie73 `x[ , 1]` sounds fine. just a problem with my editing in-browser.

sritchie73 commented:
@MichaelChirico i've updated the code and tests so that `rownames = FALSE` no longer generates a warning. I've also removed the similar warnings generated when `rownames = NULL` or `rownames = NA`.

MichaelChirico commented:
LGTM
